metals/src/main/scala/scala/meta/internal/metals/codeactions/ActionableDiagnostic.scala

```diff
@@ -0,0 +1,65 @@
+package scala.meta.internal.metals.codeactions
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals._
+import scala.meta.pc.CancelToken
+
+import org.eclipse.{lsp4j => l}
+
+class ActionableDiagnostic() extends CodeAction {
+  override def kind: String = l.CodeActionKind.QuickFix
+
+  override def contribute(
+      params: l.CodeActionParams,
+      token: CancelToken,
+  )(implicit ec: ExecutionContext): Future[Seq[l.CodeAction]] = {
+
+    def path =
+      params.getTextDocument.getUri.toAbsolutePath.toURI.toString
+
+    def createDiagnosticAction(
+        diagnostic: l.Diagnostic,
+        relatedInformation: l.DiagnosticRelatedInformation,
+    ): l.CodeAction = {
+      val codeAction = new l.CodeAction()
+      val to = relatedInformation.getMessage
+      codeAction.setKind(l.CodeActionKind.QuickFix)
+      codeAction.setTitle(ActionableDiagnostic.title)
+      codeAction.setDiagnostics(List(diagnostic).asJava)
+      val range = relatedInformation.getLocation.getRange
+      val textEdit = new l.TextEdit(range, to)
+      codeAction.setEdit(
+        new l.WorkspaceEdit(
+          Map(path -> List(textEdit).asJava).asJava
+        )
+      )
+      codeAction
+    }
+
+    val codeActions = params
+      .getContext()
+      .getDiagnostics()
+      .asScala
+      .groupBy {
+        case ScalacDiagnostic.ScalaAction(relatedInformation) =>
+          Some(relatedInformation)
+        case _ => None
+      }
+      .collect {
+        case (Some(relatedInformation), diags)
+            if params.getRange().overlapsWith(diags.head.getRange()) =>
+          createDiagnosticAction(diags.head, relatedInformation)
```

Comment:
I think I'm a bit confused by this, but I might be missing it. Do you by chance have a trace that you could show from scala-cli about what this full diagnostic looks like with the related information?

If I've following this right, this assumes that related information is actionable right? But by definition, diagnostic related information isn't alway though right? Like in the situation where you have a diagnostic that also has a related info about another part of code that it relates to doesn't have any type of action tied to it, so this should't provide a code action in that case. Am I misunderstanding it, or wouldn't this provide a code action for every diagnostic that has a related information that has the same position as the diagnostic?

Reply:
It's kind of a hack really right now, that's why we first enabled it for ScalaCLI, since there is no way of defining what should be done with the related diagnostics. It would actually be great to be able get the `kind` of diagnostics :thinking:

Reply:
Ahhh, I'd highly recommend not going against the spec for this then. This will sort of break assumptions then since the spec is pretty well defined here on how this should work. We haven't really used them at all, but if I'm understanding the flow correctly instead here we should be using `data`, which is part of the `Diagnostic` and can be persisted between a `textDocument/publishDiagnostics` and a `textDocument/codeAction`. This way when someone would trigger a code action on that diagnostic, the related information has nothing to do with the fix, since it's not meant to. We could then check against the data to know we can apply what is needed, and then apply it. If i'm not mistaken clang and tsserver do just this, they actual contain the fix right in the diagnostic data, so if it's triggered it can immediately be applied. I believe this is the approach we should follow instead of hacking this in this way.

Reply:
Makes sense, but we would need to have the same thing for BSP's diagnostic, which is currently lacking. We only have RelatedInformation. We should add an object there then I think, though it might be ok for ScalaCLI use RelatedInformation for the time being :thinking:  Maybe for BSP we could actually add a specific `EditSuggestion`?

Reply:
I think I'm still confused by the entire flow here. Do you have chance have a trace that I could look at going from diagnostic -> code action so I can see all the steps? Like where is the actual fix included? If it's just part of the related information, then I don't think that's where it should be as this is going against the [spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnosticRelatedInformation) and would need to be changed later the moment we start using this the correct way.

> Represents a related message and source code location for a diagnostic.
> This should be used to point to code locations that cause or are related to
> a diagnostics, e.g when duplicating a symbol in a scope.

For example this should be used just for extra information about other parts of your code, that's why by default it has a location so that location can be navigated to like you see [here](https://contributors.scala-lang.org/t/revisiting-dotty-diagnostics-for-tooling/5649/12?u=ckipp01).

Reply:
Sure, that was the main hacky part, which is why I wanted to get someone to double check the decision :sweat_smile: 

I think we could use data, but despite `where Diagnostic is defined as it is in LSP.` in https://build-server-protocol.github.io/docs/specification.html#publish-diagnostics it's not actually true and we don't have an additional `data` object.

@lwronski would you be able to add that field to BSP, we could use it to be more correct here.

Reply:
@tgodzik @ckipp01 Sorry for the delay in answering, meanwhile I added a `data` field in `Diagnostic` in `bsp4j` - https://github.com/build-server-protocol/build-server-protocol/pull/363.

Instead of using the `DiagnosticRelatedInformation` field to send actionable actions, I use `data` field in `Diagnostic`.  In ScalaCLI I create `textEdit` with actionable action and then serialized object and save into `data` field.

Reply:
Very cool @lwronski! I'll take a look. Do you have a trace of this that I could look at to see what the entire flow looks like? I'm assuming the latest release of scala-cli isn't sending these yet?

Reply:
Scala CLI will send the following diagnostic, but yes, it will work after release to scala-cli `0.1.17`. In latest release is missing field `annotationId` in `TextEdit`.

```
"diagnostics": [
    {
      "range": {
        "start": {
          "line": 0,
          "character": 15
        },
        "end": {
          "line": 0,
          "character": 40
        }
      },
      "severity": 4,
      "source": "scala-cli",
      "message": "com.lihaoyi::os-lib:0.7.8 is outdated, update to 0.8.1\n     com.lihaoyi::os-lib:0.7.8 -\u003e com.lihaoyi::os-lib:0.8.1",
      "data": {
        "range": {
          "start": {
            "line": 0,
            "character": 15
          },
          "end": {
            "line": 0,
            "character": 40
          }
        },
        "newText": "com.lihaoyi::os-lib:0.8.1",
        "annotationId": "update dependency to 0.8.1"
      }
    }
  ]
```

For now, I disable actionable diagnostic by default because it is still in progress and you have to pass flag. But when it will be ready, I think we can enable it by default.